perl/modules/Zulip/lib/BarnOwl/Module/Zulip.pm

```diff
@@ -0,0 +1,522 @@
+use AnyEvent::TLS;
+use JSON;
+use MIME::Base64;
+use URI;
+use URI::Encode;
+use BarnOwl::Hooks;
+use BarnOwl::Message::Zulip;
+use HTTP::Request::Common;
+use Getopt::Long qw(GetOptionsFromArray);
+
+sub fail {
+    my $msg = shift;
+    undef %cfg;
+    undef $queue_id;
+    undef $last_event_id;
+
+    BarnOwl::admin_message('Zulip Error', $msg);
+    die("Zulip Error: $msg\n");
+}
+
+
+sub initialize {
+    my $conffile = BarnOwl::get_config_dir() . "/zulip";
+
+    if (open(my $fh, "<", "$conffile")) {
+        read_config($fh);
+        close($fh);
+    }
+    if ($cfg{'api_url'} =~ /^https/) {
+        if (exists $cfg{'ssl_key_file'}) {
+            $tls_ctx = new AnyEvent::TLS(verify => $cfg{'ssl_verify'}, 
+                                         sslv3 => 0, verify_peername => "http",
+                                         ca_file => $cfg{'ssl_ca_file'},
+                                         cert_file => $cfg{'ssl_cert_file'},
+                                         key_file => $cfg{'ssl_key_file'});
+        } else {
+            $tls_ctx = new AnyEvent::TLS(verify => $cfg{'ssl_verify'}, 
+                                         sslv3 => 0, verify_peername => "http",
+                                         ca_file => $cfg{'ssl_ca_file'});
+        }           
+    } else {
+        # we still want it for a unique id
+        $tls_ctx = int(rand(1024));
+    }
+}
+
+
+sub authorization {
+    return "Basic " . encode_base64($cfg{'user'} . ':' . $cfg{'apikey'}, "")
+}
+
+sub read_config {
+    my $fh = shift;
+
+    my $raw_cfg = do {local $/; <$fh>};
+    close($fh);
+    eval {
+        $raw_cfg = from_json($raw_cfg);
+    };
+    if($@) {
+        fail("Unable to parse config file: $@");
+    }
+
+    if(! exists $raw_cfg->{user}) {
+        fail("Account has no username set.");
+    }
+    my $user = $raw_cfg->{user};
+    if(! exists $raw_cfg->{apikey}) {
+        fail("Account has no api key set.");
+    }
+    my $apikey = $raw_cfg->{apikey};
+    if(! exists $raw_cfg->{api_url}) {
+        fail("Account has no API url set.");
+    }
+    my $api_url = $raw_cfg->{api_url};
+
+    if(! exists $raw_cfg->{default_realm}) {
+        fail("Account has no default realm set.");
+    }
+    my $default_realm = $raw_cfg->{default_realm};
+
+    if( exists $raw_cfg->{ssl}) {
+        # mandatory parameters
+        if (! exists $raw_cfg->{ssl}->{ca_file}) {
+            fail("SSL parameters specified, but no CA file set");
+        }
+        $cfg{'ssl_ca_file'} = $raw_cfg->{ssl}->{ca_file};
+        $cfg{'ssl_verify'} = 1;
+        # optional parameters
+        if ( (exists $raw_cfg->{ssl}->{cert_file}) && exists $raw_cfg->{ssl}->{key_file}) {
+            $cfg{'ssl_cert_file'} = $raw_cfg->{ssl}->{cert_file};
+            $cfg{'ssl_key_file'} = $raw_cfg->{ssl}->{key_file};
+        }  else {
+            warn "SSL parameters specified, but no client credentials set.";
+        }
+    } else {
+	$cfg{'ssl_verify'} = 0;
```

Comment:
No.

This is a really bad default in 2017, even with the warning (and even for someone who understands why they shouldn’t ignore the warning: the consequence of making some typo in the config file shouldn’t be that your API key gets exposed). If a user really needs to skip SSL verification, they should be explicit about it.

Reply:
You are, of course, correct. I'm rewriting this, especially in light of the belated discovery that with no CA file specified, AnyEvent will use a platform-dependent default.